doc/mgmt/gnmi/master_arbitration.md

```diff
@@ -0,0 +1,331 @@
+
+
+A `MasterArbitration` message is required in the `SetRequest` if Master Arbitration is enabled. `SET` RPC handler is expanded to use one of two functions that handle incoming `SetRequest` differently if the Master Arbitration is enabled or disabled. By default, disabling Master Arbitration is applied:
+
+```
+// ReqFromMasterDisabledMA always returns true. It is used when Master Arbitration
+// is disabled.
+func ReqFromMasterDisabledMA(req *gnmipb.SetRequest) bool {
+  return true
+}
+
+// ReqFromMaster point to a function that is called to verify if the request comes
+// from a master controller.
+ReqFromMaster := ReqFromMasterDisabledMA
+
+func (s *Server) Set(ctx context.Context, req *gnmipb.SetRequest) (*gnmipb.SetResponse, error) {
+  if !reqFromMaster(req) {
+    return nil, status.Error(codes.PermissionDenied, "Not a master")
+  }
+
+  // ...
+
+```
+
+If Master Arbitration is enabled, the following check is executed:
+
+```
+type uint128 struct {
+  High uint64
+  Low uint64
+}
+
+func (lh *uint128) Compare(rh *uint128) {
+  if lh.High > rh.High {
+    return 1;
+  }
+  if lh.High < rh.High {
+    return -1;
+  }
+  if lh.Low > rh.Low {
+    return 1;
+  }
+  if lh.Low < rh.Low {
+    return -1;
+  }
+  return 0;
+}
+
+var masterEID = uint128{High: 0, Low: 0}
+
+// ReqFromMasterEnabledMA returns true if the request is sent by the master
+// controller.
+func ReqFromMasterEnabledMA(req *gnmipb.SetRequest) bool {
+  // Read the election_id.
+  reqEID := uint128{High: 0, Low: 0}
+  // It can be one of many extensions, so iterate through them to find it.
+  for _, e := range req.extension {
+    ma := e.GetMasterArbitration()
+    if ma == nil {
+      continue
+    }
+    // The Master Arbitration descriptor has been found.
+    reqEID = uint128{High: ma.ElectionID.High, Low = ma.ElectionID.Low}
+    // Use the election ID that is in the last extension, so, no 'break' here.
+  }
+  switch masterEID.Compare(reqEID) {
+  case 1:  // This Election ID is smaller than the known Master Election ID.
+    return false
+  case -1: // New Master Election ID received!
+    masterEID = reqEID
+  }
+  return true
+}
+```
+
+#### Manifest (if the feature is an Application Extension)
+
+N/A
+
+#### CLI/YANG model Enhancements
+
+N/A
+
+#### Config DB Enhancements
+
+N/A
+
+### Warmboot and Fastboot Design Impact
+
+N/A
+
+### Restrictions/Limitations
+
+Master Arbitration is designed to enable one master client for one role. However, only default role is supported for now.  In other words, the content of the role field is ignored.
+
+### Testing Requirements/Design
+The main focus of the test is on the coverage of the new feature. The user cases are listed as following:
+
+* No Master Arbitration extension protobuf when Master Arbitration is disabled
+  * Nothing should happen
```

Comment:
**Regarding the issue, we discussed in community-meeting to avoid issues with EID out-of-sync between new-controller-master & GNMI-server till new SET from user through new controller:**  ==> I was also suggesting to learn new EID from get-request from new-master rather than waiting for set initiated by user through new-master

* **While checking further, this GNMI spec** (https://github.com/openconfig/reference/blob/master/rpc/gnmi/gnmi-master-arbitration.md#32-setrequest-handling), there is a special mention about this case to sync latest EID immediately through empty-set rather than waiting for set-from-user, after new master change. 

**Exact snippet from spec:**
![image](https://github.com/sonic-net/SONiC/assets/55006784/8fb4bcd4-2f5b-4f75-8b14-640d26b16454)

Assuming controller on master change will immediately does empty-set with new EID. So from SONIC GNMI-server perspective, hope we will handle empty-GNMI-set and update latest EID through that, and ignore the subsequent sets from old-master. 

Can you please confirm this EID-sync handling in GNMI-server for empty-sets?

Reply:
Yes. EID check is done for every SET request (for both empty-set and non-empty-set). As long as new master controller sending empty-set, GNMI-server will sync the master EID.

Reply:
ok thanks.